src/inc/til/u8u16convert.h

```diff
@@ -306,25 +89,117 @@ namespace til // Terminal Implementation Library. Also: "Today I Learned"
+            {
+                if (state.want > len8) // we still didn't get enough data to complete the code point, however this is not an error
+                {
+                    std::move(cursor8, cursor8 + len8, state.partials + state.have);
+                    state.have += gsl::narrow_cast<uint8_t>(len8);
+                    state.want -= gsl::narrow_cast<uint8_t>(len8);
+                    out.clear();
+                    return S_OK;
+                }
+
+                std::move(cursor8, cursor8 + state.want, state.partials + state.have);
+                len16 = MultiByteToWideChar(CP_UTF8, 0UL, state.partials, gsl::narrow_cast<int>(state.have) + state.want, out.data(), capa16);
+                RETURN_HR_IF(E_UNEXPECTED, !len16);
+
+                capa16 -= len16;
+                len8 -= state.want;
+                cursor8 += state.want;
+                state.reset();
+            }
+
+            if (len8)
+            {
+                auto backIter{ cursor8 + len8 };
+                // If the last byte in the string was a byte belonging to a UTF-8 multi-byte character
+                if (backIter[-1] & 0b1'0000000)
+                {
+                    // Check only up to 3 last bytes, if no Lead Byte is found then the byte before will be the Lead Byte and no partials are in the string
+                    const auto stopLen{ std::min(len8, 3) };
+                    for (auto sequenceLen{ 1 }; sequenceLen <= stopLen; ++sequenceLen)
+                    {
+                        --backIter;
+                        // we found a Lead Byte if at least 2 high-order bits are set
+                        if ((*backIter & 0b11'000000) == 0b11'000000)
+                        {
+                            // determine the number of bytes of the code point indicated by the lead byte
+                            const auto codePointLen{
+                                2 + // the code point consists of at least 2 bytes because we found a lead byte
+                                gsl::narrow_cast<int>((*backIter & 0b111'00000) == 0b111'00000) + // add one if at least 3 high-order bits of the lead byte are set
+                                gsl::narrow_cast<int>((*backIter & 0b1111'0000) == 0b1111'0000) // add another one if 4 high-order bits of the lead byte are set
+                            };
+
+                            // if the length of the code point differs from the length we stepped backward, we will capture the last sequence which is a partial code point
+                            if (codePointLen != sequenceLen)
+                            {
+                                std::move(backIter, backIter + sequenceLen, state.partials);
+                                len8 -= sequenceLen;
+                                state.have = gsl::narrow_cast<uint8_t>(sequenceLen);
+                                state.want = gsl::narrow_cast<uint8_t>(codePointLen - sequenceLen);

```

Comment:
This can result in 255 for invalid UTF8 encodings, because codePointLen can be greater than sequenceLen, right?

Reply:
Fortunately this can't overflow. codePointLen is expected to be greater than sequenceLen, right. But we subtract sequenceLen from codePointLen here.

Reply:
I think we're best served if we use [Chris Wellons' branchless utf8 decoder](https://github.com/skeeto/branchless-utf8/blob/master/utf8.h) as a template here (while crediting him appropriately of course).
This would result in this simpler and faster algorithm:
```cpp
constexpr uint8_t utf8width(char ch) {
    static constexpr uint8_t lengths[] = {1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 2, 2, 2, 2, 3, 3, 4, 0};
    return lengths[ch >> 4];
}
```
Afterwards we can check if the result is 0 and reject the codepoint.

Reply:
Oh sorry I meant the inverse: `sequenceLen` being greater than `codePointLen`.
As far as I can see this can happen with something like `"foo\xCBar"` which is
```
... 1100'1011 0110'0001 0111'0010
      \xCB        a         r
```

`sequenceLen` should be 3 at this point and `\xCB` will equal to a `codePointLen` of 2.

Reply:
For this particular example nothing is gonna be wrong because we wouldn't even iterate backwards if we see ASCII at the end of the string. However, I understand your concern. The comparison should have been
`if (codePointLen > sequenceLen)`
If the lead byte indicates 2 bytes but we iterated already through 3 bytes then we got invalid UTF-8. In this case nothing is to be captured and the invalid sequence is to be converted into replacement characters.